src/Build/Evaluation/Expander.cs

```diff
@@ -1089,6 +1107,15 @@ private static class PropertyExpander<T>
                             propertyValue = LookupProperty(properties, expression, propertyStartIndex + 2, propertyEndIndex - 1, elementLocation, usedUninitializedProperties);
                         }
 
+                        if (truncate && propertyValue != null)
+                        {
+                            var value = propertyValue.ToString();
+                            if (value.Length > DefaultTruncationCharacterLimit)
+                            {
+                                propertyValue = value.Substring(0, DefaultTruncationCharacterLimit) + "...";
```

Comment:
This PR truncates individual values to 1024, not the entire message, so you might still have:

    Task "Foo" skipped, due to false condition; ('@(MyItems)' != '' and '$(Bar)' != 'true') was evaluated as ('Some;Really;Long;String;...' != '' and '' != 'true')

I could truncate the entire `Condition` message instead, but that might make the log message a lot less readable? I'm open to other ideas, though.

As for truncating property values sooner, it seems like properties are created with the values expanded already:

https://github.com/dotnet/msbuild/blob/116af13e6760ebbb8466174201f1ebbc8df11dfa/src/Build/Instance/ProjectPropertyInstance.cs#L170

There doesn't appear to be an opportunity to shorten any earlier? It would still be a `Substring()` call?

Reply:
> This PR truncates individual values to 1024, not the entire message, so you might still have:
> 
Right; that's what I meant. One option (again, making it more complicated, so maybe not worth it) is to let the whole string be added to result, then when we get to concatenating them, look at how long that would make the whole string. Then you can truncate as necessary to keep the _total_ length under 1024. Since you'd have to be aware of empty strings, that would make it noticeably more complicated—to be honest, I think I'd be just as happy with changing DefaultTruncationCharacterLimit to something that indicates that you might get an output longer than that like MaxPropertyLength.

> As for truncating property values sooner, it seems like properties are created with the values expanded already:
> 
> https://github.com/dotnet/msbuild/blob/116af13e6760ebbb8466174201f1ebbc8df11dfa/src/Build/Instance/ProjectPropertyInstance.cs#L170
> 
> There doesn't appear to be an opportunity to shorten any earlier? It would still be a `Substring()` call?

Never mind; you're right. I wish we could get it as a string rather than an object, too, but I don't think that's particularly doable. I tried to refactor all of this into spans shortly after I started but got bogged down in places where I had to interface with the outside world, unfortunately.

Reply:
> I think I'd be just as happy with changing DefaultTruncationCharacterLimit to something that indicates that you might get an output longer than that like MaxPropertyLength.

Should I change the value of `DefaultTruncationCharacterLimit`? I looked for a `MaxPropertyLength` in the codebase, but didn't find it.

Or should I make the `<summary/>` of `DefaultTruncationCharacterLimit` more descriptive?

Reply:
Sorry—that wasn't clear. MaxPropertyLength doesn't exist; I was thinking of changing the name of DefaultTruncationCharacterLimit to something else, so essentially making the summary more descriptive but starting with the name. I don't know that MaxPropertyLength is right either; I'm just going for something that indicates that the total length might be longer than this.

Reply:
They are now named `CharacterLimitPerExpansion` and `ItemLimitPerExpansion`, I think these are ok names?

Reply:
Yep! I like that a lot better.